accepted/2020/serializer/WriteableDomAndDynamic.md

```diff
@@ -0,0 +1,587 @@
+
+        // Support terse syntax for JsonArray (no cast necessary to JsonArray).
+        public System.Text.Json.Serialization.JsonNode? this[int index] { get; set; }
+
+        // Support terse syntax for JsonObject (no cast necessary to JsonObject).
+        public virtual System.Text.Json.Serialization.JsonNode? this[string key] { get; set; }
+
+        // Support terse syntax for JsonValue (no cast necessary to JsonValue to change a value).
+        // This represents the internal value, and no conversion is performed.
+        public virtual object Value {get; set;}
+
+        // Return the internal value or convert to T as necessary.
+        public abstract T GetValue<T>();
+        public abstract object GetValue<Type type>();
+        public abstract bool TryGetValue<T>(out T? value);
+        public abstract bool TryGetValue(Type type, out object? value);
+
+        // Convert a non-dynamic node to a dynamic.
+        public object ToDynamic();
+        // This would also work assume the linker will remove the reference to SLE if not called:
+        // public dynamic ToDynamic();
+
+        // The token type from deserialization; otherwise JsonValueKind.Unspecified.
+        public JsonValueKind ValueKind { get; }
+
+        // Serialize() wrappers which pass the value and options.
+        // These are helpers and thus can be considered an optional API feature.
+        // There are no wrappers for JsonSerializer.Deserialize* methods since the
+        // existing Deserialize* methods can be called.
+        public string Serialize();
+        public void Serialize(System.Text.Json.Utf8JsonWriter writer);
+        public System.Threading.Tasks.Task SerializeAsync(Stream utf8Json, CancellationToken cancellationToken = default);
+        public byte[] SerializeToUtf8Bytes();
+    }
+
+    public sealed class JsonArray : JsonNode, IList<object>
+    {
+        public JsonArray(JsonSerializerOptions options = null);
+
+        // Can be used to return:
+        // - Any compatible collection type
+        // - JsonElement
+        // - string or byte[] (as JSON)
+        public override T GetValue<T>();
+        public override object GetValue<Type type>();
+        public override bool TryGetValue<T>(out T? value);
+        public override bool TryGetValue(Type type, out object? value);
+    }
+
+    public sealed class JsonObject : JsonNode, IDictionary<string, object>
+    {
+        public JsonObject(JsonSerializerOptions options = null);
+
+        // Can be used to return:
+        // - Any compatible dictionary type
+        // - POCO including anonymous type or record
+        // - JsonElement
+        // - string or byte[] (as JSON)
+        public override T GetValue<T>();
+        public override object GetValue<Type type>();
+        public override bool TryGetValue<T>(out T? value);
+        public override bool TryGetValue(Type type, out object? value);
+    }
+
+    public sealed class JsonValue : JsonNode
```

Comment:
If this type existed in isolation, I can see a case to be made for it, but I suspect given the pervasive use of the union type JsonElement, having another very similar concept which represents a list of JSON values, would cause user confusion.
That was one of the benefits of keeping explicit JsonString, JsonNumber, JsonNull and JsonBoolean types from the initial prototype.